### PR TITLE
Update validator shuffling

### PR DESCRIPTION
**Pull request by drozdziak1**
# Update validator shuffling
This PR updates and improves validator shuffling generation.

This most likely needs some polishing. There's a couple things I'm not happy
with yet:
* [x] The new epoch choice is too big - and therefore surrounded only by
  the active validators who didn't leave the set yet (no "soon-to-leaves" useful
  for delay edge case testing)
* [x] I did not consider the delays outlined by the spec

## Original commit message (may change as I squash new things into it):

Update test vec generation and make it dynamic

This commit updates the validator shuffling generator and tries to make
the test validator sets more life-like.

constants.py:
* Add FAR_FUTURE_EPOCH

core_helpers.py:
* Update the helpers needed for validator shuffling

Remove enums.py: status flags are not used since 2018-12-28

tgen_shuffling.py:
* Use a randomized probabilistic approach for generating configurably
diverse validator sets; pre-FAR_FUTURE_EPOCH exit epochs truncated for
readability
* Change the YAML layout to include validator data needed for
present-day shuffling

yaml_objects.py:
* Change ValidatorRecord name to Validator (as in spec); change fields
for up-to-date shuffling
* Remove ShardCommittee (not used for shuffling at all as of today)

test_vector_shuffling.yml:
* Regenerate

### CODE REVIEW

File: eth2_testgen/shuffling/core_helpers.py [hunk 1]
```diff
@@ -8,26 +8,29 @@
...
 
 from constants import EPOCH_LENGTH, SHARD_COUNT, TARGET_COMMITTEE_SIZE
-from enums import ValidatorStatusCode
 from utils import hash
-from yaml_objects import ShardCommittee, ValidatorRecord
+from yaml_objects import Validator
 
+EpochNumber = int
+ValidatorIndex = int
+Bytes32 = bytes
```
Comment (jannikluhn) [inline eth2_testgen/shuffling/core_helpers.py]: These three should be `EpochNumber = NewType("EpochNumber", int)` etc.

File: eth2_testgen/shuffling/core_helpers.py [hunk 2]
```diff
@@ -78,57 +81,51 @@ def shuffle(values: List[Any], seed: Hash32) -> List[Any]:
...
+    committee is itself a list of validator indices.
+    """
+
+    active_validator_indices = get_active_validator_indices(validators, epoch)
+
+    committees_per_epoch = get_epoch_committee_count(
+        len(active_validator_indices))
+
+    # Shuffle
+    seed = xor(seed, int_to_bytes32(epoch))
```
Comment (jannikluhn) [inline eth2_testgen/shuffling/core_helpers.py]: Just a note for others because I got confused: This line has been removed in the spec, but it's still in the January release which I guess we want to test against right now.

File: eth2_testgen/shuffling/tgen_shuffling.py [hunk 3]
```diff
@@ -39,24 +41,50 @@ def yaml_ValidatorStatusCode(dumper, data):
...
-                original_index=num_val)
-            for num_val in range(num_val)
-        ]
+        idx_max = random.randint(128, 512)
+
+        validators = []
+        for idx in range(idx_max):
+            v = Validator(original_index=idx)
+            # 4/5 of all validators are active
+            if random.randint(0, 4):
```
Comment (jannikluhn) [inline eth2_testgen/shuffling/tgen_shuffling.py]: I think it would be a bit cleaner to use `random.random() < 0.8`, because it also allows fractions other than `(n - 1) / n`

Comment (jannikluhn) [reply eth2_testgen/shuffling/tgen_shuffling.py]: Also, I think this would be easier to understand without all the nesting. So I suggest to pick a random number in the beginning, and then do something like this:

```Py
if r < 0.2:
    v.activation_epoch = FAR_FUTURE_EPOCH
    v.exit_epoch = FAR_FUTURE_EPOCH
elif r < 0.2 + 0.25:
    # validator will exit in foreseeable future
elif r < 0.45 + 0.25:
    # ...
```

(even though adding the probabilities feels a bit messy, there's probably a better pattern for this)

Comment (drozdziak1) [reply eth2_testgen/shuffling/tgen_shuffling.py]: I really like the floating number approach. 

I'm not really sold on the `elif` with probability addition though. I don't think three-deep ifs are as scary.

File: eth2_testgen/shuffling/core_helpers.py (hunk 2 above)
Comment (drozdziak1) [reply eth2_testgen/shuffling/core_helpers.py]: Copied everything 1-to-1 from `v0.1`, mentioned the release link at the top of the file.

File: eth2_testgen/shuffling/core_helpers.py (hunk 1 above)
Comment (drozdziak1) [reply eth2_testgen/shuffling/core_helpers.py]: Done

**Comment by drozdziak1** [comment]
FYI this is up-to-date with v0.1 now. Applied most of the advice from @jannikluhn
